Modules/Loadable/Markups/MRMLDM/vtkMRMLMarkupsDisplayableManager.cxx

```diff
@@ -939,43 +916,23 @@ vtkSlicerMarkupsWidget * vtkMRMLMarkupsDisplayableManager::CreateWidget(vtkMRMLM
     {
     return nullptr;
     }
-  vtkMRMLAbstractViewNode* viewNode = vtkMRMLAbstractViewNode::SafeDownCast(this->GetMRMLDisplayableNode());
-  vtkRenderer* renderer = this->GetRenderer();
 
-  vtkSlicerMarkupsWidget* widget = nullptr;
-  if (vtkMRMLMarkupsAngleNode::SafeDownCast(markupsNode))
-    {
-    widget = vtkSlicerAngleWidget::New();
-    }
-  else if (vtkMRMLMarkupsFiducialNode::SafeDownCast(markupsNode))
-    {
-    widget = vtkSlicerPointsWidget::New();
-    }
-  else if (vtkMRMLMarkupsLineNode::SafeDownCast(markupsNode))
-    {
-    widget = vtkSlicerLineWidget::New();
-    }
-  else if (vtkMRMLMarkupsCurveNode::SafeDownCast(markupsNode))
-    {
-    widget = vtkSlicerCurveWidget::New();
-    }
-  else if (vtkMRMLMarkupsClosedCurveNode::SafeDownCast(markupsNode))
-    {
-    widget = vtkSlicerClosedCurveWidget::New();
-    }
-  else if (vtkMRMLMarkupsPlaneNode::SafeDownCast(markupsNode))
-    {
-    widget = vtkSlicerPlaneWidget::New();
-    }
-  else
+  // Create a widget of the associated type if the node matches the registered nodes
+  vtkSlicerMarkupsWidget* widget =
+    vtkSlicerMarkupsWidget::SafeDownCast(
+      vtkMRMLMarkupsRegistrationFactory::GetInstance()->GetWidgetByMarkupsNodeClass(markupsNode->GetClassName())
+    )->CreateInstance();
```

Comment:
```suggestion
  vtkSlicerMarkupsWidget* widget =
    vtkSlicerMarkupsWidget::SafeDownCast(
      this->GetMRMLApplicationLogic()->GetWidgetByMarkupsNodeClass(markupsNode->GetClassName())
    )->CreateInstance();
```

Reply:
The advantage of adding the registration logic to `vtkMRMLApplicationLogic` is that we can:
* avoid using our own factory and rely on the fact there a single instance of the application
* support access to function related to widget registration from displayable manager (since they derive from `vtkMRMLAbstractLogic` having the  method [GetMRMLApplicationLogic](https://github.com/Slicer/Slicer/blob/87adcd2211d69826b5bdba8508b831ae4b919295/Libs/MRML/Logic/vtkMRMLAbstractLogic.h#L145-L146))

Using `vtkSlicerMarkupsLogic` could be confusing as we can have multiple instance of the logic

Reply:
>  forced to use abstract types for the registration

This could be addressed by moving `vtkMRMLMarkupsNode` and `vtkMRMLMarkupsJsonStorageNode` to `Libs/MRML/Core`  (`vtkMRMLMarkupsStorageNode` is already there)